torch/distributed/_functional_collectives_impl.py

```diff
@@ -127,7 +127,19 @@ def _register_tensor_wrapper(tensor) -> None:
         # Tensor storage -> work mapping is maintained in C++
         return
     global data_ptr_to_work
-    data_ptr = tensor.elem.data_ptr()
+
+    # FIXME: This is almost definitely a bug.
+    if isinstance(
+        tensor.elem,
+        (
+            torch._subclasses.fake_tensor.FakeTensor,
+            torch._subclasses.functional_tensor.FunctionalTensor,
+        ),
+    ):
+        data_ptr = 0
+    else:
+        data_ptr = tensor.elem.data_ptr()
```

Comment:
cc @yifuwang

Reply:
Yeah this data_ptr -> work map getting filled in when we trace through `AsyncCollectiveTensor` during PT2 tracing is almost definitely one of the reasons `AsyncCollectiveTensor` (supposedly) doesn't work great with `torch.compile` today at the graph boundaries. I told @wanchaol I would try to take a look at this one some other DTensor work wraps up, although probably not in the next month or two

Reply:
Here as well, should this be a WeakStorageIdDict ?

Reply:
FYI this code has been deprecated and I'm planning to remove it soon.

The tensor -> work tracking has been moved to C++ and the registry uses `getWeakStorageImpl()` as the key.

https://github.com/pytorch/pytorch/blob/e61aaab72562fe95d76e77b01678d71467e1739e/torch/csrc/distributed/c10d/Functional.cpp#L15-L24